### PR TITLE
Don't generate new tags for column type changes

### PR DESCRIPTION
**Pull request by fulghum**
Currently, when a column's type changes to a new type family (e.g. changing from `varchar` to `int`), we assign a new column tag. This prevents us from easily identifying the column across type changes. For example:
* If column X is renamed and then its type is later changed, we can't track it across the rename or type change anymore. This usage pattern removes the value that column tags were intended to provide. 
* If column X changes its type, we can't easily differentiate between column X being dropped and a new column X being added. Heuristically, we can look at all the data in the table if it's all exactly identical, then we know it's a rename, but this is a somewhat hacky heuristic that can become ambiguous in more nuanced cases. This currently causes us to generate incorrect SQL patch statements that drops the renamed column and adds a new column. If customers were to apply these incorrect patch statements, it would drop all the data from their existing column, instead of converting it to the new type. This affects `dolt_patch()` and `dolt diff -r sql` and also prevents us from generating correct DDL statements for binlog replication.

This PR changes schema alterations so that columns now retain their original column tag across type changes, which fixes the issues above. 

A side effect of this change is that if a customer is working on a branch, creates a column, and changes its type, the tag won't match if the schema on another branch is updated to the final type change (i.e. without going through the initial/intermediary type). Code was originally added for this case (see: https://github.com/dolthub/dolt/issues/3950), however, since then, we have added support for schema merging and this isn't an issue anymore. I've confirmed that we can successfully merge non-matching column tags, from both directions, and added a test for this case, too.

Longer-term, I agree we should continue exploring removing tags completely, but in the short-term, this fixes a correctness problem with SQL patch statement generation that binlog support needs. 

Related to: https://github.com/dolthub/dolt/issues/6710#issuecomment-1734318087

### CODE REVIEW

**Comment by coffeegoddd** [comment]
@fulghum __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>e459ede</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>e459ede</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Review by zachmu — APPROVED** [review]
LGTM

**Comment by coffeegoddd** [comment]
@fulghum __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>d12cdec</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>d12cdec</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

File: go/libraries/doltcore/sqle/enginetest/dolt_queries.go [hunk 1]
```diff
@@ -2072,11 +2072,16 @@ var HistorySystemTableScriptTests = []queries.ScriptTest{
...
+				// tags have changed in previous versions of the table, the index tags won't match up completely.
 				// https://github.com/dolthub/dolt/issues/6891
+				// NOTE: {4,5,nil} shows up as a row from the first commit, when c2 was a varchar type. The schema
+				//       for dolt_history_t uses the current table schema, and we can't extract an int from the older
+				//       version's tuple, so it shows up as a NULL. In the future, we could use a different tuple
+				//       descriptor based on the version of the row and pull the data out that way and try to convert
+				//       it to the new type, but it may not actually be worth it.
 				Query:    "select pk, c1, c2 from dolt_history_t where pk=4;",
-				Expected: []sql.Row{{4, 5, 6}},
+				Expected: []sql.Row{{4, 5, 6}, {4, 5, nil}},
```
[the remark this answers is not in the document]
Comment (fulghum) [reply go/libraries/doltcore/sqle/enginetest/dolt_queries.go:2084]: Yeah, seems like a good place for a session warning. I'll go ahead and merge the PR in, and will follow up this afternoon to see if I can cleanly get a warning issued.

[the remark this answers is not in the document]
Comment (fulghum) [reply go/libraries/doltcore/sqle/enginetest/dolt_queries.go:2084]: Added in https://github.com/dolthub/dolt/pull/7776

**Pull request merged**